arangod/RocksDBEngine/RocksDBChecksumEnv.cpp

```diff
@@ -0,0 +1,264 @@
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Julia Puget
+////////////////////////////////////////////////////////////////////////////////
+
+#include "RocksDBChecksumEnv.h"
+
+#include "Basics/Exceptions.h"
+#include "Basics/files.h"
+#include "Basics/FileUtils.h"
+#include "Basics/MutexLocker.h"
+#include "Logger/LogMacros.h"
+
+namespace arangodb::checksum {
+
+ChecksumCalculator::ChecksumCalculator()
+    :
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+      _context(EVP_MD_CTX_new()) {
+#else
+      _context(EVP_MD_CTX_create()) {
+#endif
+  if (_context == nullptr) {
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_OUT_OF_MEMORY);
+  }
+  if (EVP_DigestInit_ex(_context, EVP_sha256(), nullptr) == 0) {
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL,
+                                   "unable to initialize SHA256 processor");
+  }
+}
+
+void ChecksumCalculator::computeFinalChecksum() {
+  TRI_ASSERT(_context != nullptr);
+  unsigned char hash[EVP_MAX_MD_SIZE];
+  unsigned int lengthOfHash = 0;
+  if (EVP_DigestFinal_ex(_context, hash, &lengthOfHash) == 0) {
+    TRI_ASSERT(false);
+  }
+  _checksum = basics::StringUtils::encodeHex(
+      reinterpret_cast<char const*>(&hash[0]), lengthOfHash);
+}
+
+void ChecksumCalculator::updateIncrementalChecksum(char const* buffer,
+                                                   size_t n) {
+  TRI_ASSERT(_context != nullptr);
+  updateEVPWithContent(buffer, n);
+}
+
+ChecksumCalculator::~ChecksumCalculator() {
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+  EVP_MD_CTX_free(_context);
+#else
+  EVP_MD_CTX_destroy(_context);
+#endif
+}
+
+void ChecksumCalculator::updateEVPWithContent(const char* buffer, size_t n) {
+  TRI_ASSERT(_context != nullptr);
+  if (EVP_DigestUpdate(_context, static_cast<void const*>(buffer), n) == 0) {
+    TRI_ASSERT(false);
+  }
+}
+
+bool ChecksumHelper::isFileNameSst(std::string const& fileName) {
+  return TRI_Basename(fileName).size() > 4 &&
+         (fileName.compare(fileName.size() - 4, 4, ".sst") == 0);
+}
+
+bool ChecksumHelper::writeShaFile(std::string const& fileName,
+                                  std::string const& checksum) {
+  TRI_ASSERT(isFileNameSst(fileName));
+
+  std::string shaFileName = buildShaFileNameFromSst(fileName, checksum);
+
+  LOG_TOPIC("80257", DEBUG, arangodb::Logger::ENGINES)
+      << "shaCalcFile: done " << fileName << " result: " << shaFileName;
+  auto res = TRI_WriteFile(shaFileName.c_str(), "", 0);
```

Comment:
We are writing the file here directly via the OS. Maybe we should be good citizens and use the underlying ENV to create the sha file. Imagine there were another underlying env which does yet some other magic (like the proposed logging env) beneath, this underlying env would then not be aware of the additional file.

Reply:
I tend to disagree here: I think we should not use the env to write the sha files, simply because they are not RocksDB's own files. They are managed completely outside of RocksDB.

Reply:
For RocksDB's own files, we are forwarding the call to the base method. When rocksdb doesn't care about the files, like .sha., we use our own OS methods, like TRI_Unlink now.